### PR TITLE
[Concurrency] Add a unique Task ID to AsyncTask

### PR DESCRIPTION
**Pull request by fredriss**
The offset of this Id in the AsyncTask is an ABI constant. This way
introspection tools can extract the currently running task identifier
without any need for special APIs.

This PR first changes JobFlags to a 32bits entity in order to be able
to pack the ID in the freed-up space.

### CODE REVIEW

**Comment by fredriss** [comment]
@swift-ci test

**Comment by fredriss** [comment]
Obviously this changes the size of Job on 32 bits platforms:
```
/Users/buildnode/jenkins/workspace/swift-PR-macos/branch-main/swift/include/swift/ABI/Task.h:127:1: error: static_assert failed due to requirement 'sizeof(swift::Job) == 6 * sizeof(void *)' "Job size is wrong"
14:32:20 static_assert(sizeof(Job) == 6 * sizeof(void*),
14:32:20 ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@rjmccall That's expected, any preference how to relax the static assert?

**Comment by fredriss** [comment]
Windows doesn't support `std::atomic_uint32_t` ? I suppose `std::atomic<uint32_t>` will do

File: include/swift/ABI/Task.h [hunk 1]
```diff
@@ -485,13 +490,24 @@ class AsyncTask : public Job {
     return reinterpret_cast<AsyncTask *&>(
         SchedulerPrivate[NextWaitingTaskIndex]);
   }
+
+  /// Get the next non-zero Task ID.
+  uint32_t GetNextTaskId() {
```
Comment (rjmccall) [inline include/swift/ABI/Task.h:495]: Style nit: we're consistent about using `camelCase` for method names in Swift.

File: include/swift/ABI/Task.h [hunk 2]
```diff
@@ -485,13 +490,24 @@ class AsyncTask : public Job {
...
+    static std::atomic_uint32_t Id;
+    uint32_t Next = Id.fetch_add(1, std::memory_order_relaxed);
+    if (Next == 0) Next = Id.fetch_add(1, std::memory_order_relaxed);
```
Comment (rjmccall) [inline include/swift/ABI/Task.h:498]: Maybe initialize it to 1 just so we don't trip this special case the first time?  Minor, but still.

**Review by rjmccall — COMMENTED** [review]
Generally LGTM, just some slight nit-picking.

**Comment by fredriss** [comment]
@swift-ci test

**Comment by fredriss** [comment]
@swift-ci test

**Comment by fredriss** [comment]
@swift-ci Test Windows Platform

**Comment by fredriss** [comment]
This breaks 3 watch simulator tests:
```
18:36:44 Failed Tests (3):
18:36:44   Swift(watchsimulator-i386) :: Concurrency/Runtime/checked_continuation.swift
18:36:44   Swift(watchsimulator-i386) :: Concurrency/Runtime/objc_async.swift
18:36:44   Swift(watchsimulator-i386) :: Concurrency/Runtime/yielding_continuation.swift
```

The tests seem to crash. Any idea what could be wrong @rjmccall ?

**Comment by fredriss** [comment]
@swift-ci Test Linux Platform

**Comment by fredriss** [comment]
@swift-ci Test Windows Platform

**Comment by fredriss** [comment]
@swift-ci please test

**Comment by fredriss** [comment]
@swift-ci please test

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by fredriss** [comment]
@swift-ci Test macOS Platform

**Pull request merged**